### PR TITLE
Dynamic family matrix for CI jobs

### PR DESCRIPTION
**Pull request by xanthio**
In the objective of getting more contribution, these changes create the family matrix at CI run time.

I will be easier for a non CI-friendly contributor to add a new family to project as they won't have to make change in the workflow files.

Updating the Test workflow was the opportunity to update the deprecated checkout action step.

### CODE REVIEW

**Comment by xanthio** [comment]
I guess it would be a good idea to split Fetch/CMSIS/HAL in individual jobs too.

**Review by Hish15 — APPROVED** [review]

**Comment by atsju** [comment]
Indeed, extensive comments are welcome. 
I would appreciate ELI5 comments but I let your @Hish15 decide if it's needed or not. As I understand, someone speaking Unix would not need any comment.

**Comment by xanthio** [comment]
Actually its not so understandable for someone who is not a unix user.
I can split `Create matrix` step so the new steps names will comment the work that is done.

**Comment by xanthio** [comment]
In fact not easy to split as steps use list variable that github action can not handle.
That is why I was using JSON object as recommended https://docs.github.com/en/actions/learn-github-actions/expressions#example-returning-a-json-object
